File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
    static class Music implements Comparable<Music> {
        int no, plays;
        String genre;
        
        Music(int no, int plays, String genre) {
            this.no = no;
            this.plays = plays;
            this.genre = genre;
        }
        
        @Override
        public int compareTo(Music o) {
            if (o.genre.equals(this.genre)) {
                if (this.plays == o.plays) return this.no - o.no;
                return o.plays - this.plays;
            }
            return map.get(o.genre) - map.get(this.genre);
        }
    }
    
    static Map<String, Integer> map;
    
    public int[] solution(String[] genres, int[] plays) {
        map = new HashMap<>();
        for (int i = 0; i < genres.length; i++) {
            map.put(genres[i], map.getOrDefault(genres[i], 0) + plays[i]);
        }
        
        PriorityQueue<Music> q = new PriorityQueue<>();
        for (int i = 0; i < genres.length; i++) {
            q.offer(new Music(i, plays[i], genres[i]));
        }
        
        String genre = q.peek().genre;
        int cnt = 0;
        
        List<Integer> list = new ArrayList<>();
        while (!q.isEmpty()) {
            Music m = q.poll();
            cnt++;
            
            if (!genre.equals(m.genre)) {
                genre = m.genre;
                cnt = 1;
            }
            
            list.add(m.no);
            
            if (cnt == 2) {
                while (!q.isEmpty()) {
                    if (!q.peek().genre.equals(genre)) break;
                    q.poll();
                }
            }
        }
        
        int[] answer = new int[list.size()];
        for (int i = 0 ; i < list.size(); i++) {
            answer[i] = list.get(i);
        }
        
        return answer;
    }
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    const map = new Map();
    
    for (let i = 0; i < genres.length; i++) {
        map.set(genres[i], []);
    }
    
    for (let i = 0; i < genres.length; i++) {
        map.get(genres[i]).push({no: i, v: plays[i]});
    }
    
    for (let m of map.values()) {

        m.sort((a, b) => {
            if (a.v == b.v) {
                return a.no - b.no;
            }
            return b.v - a.v;
        })
    }
    
    const temp = new Map();
    
    for (let [key, value] of map) {
        let sum = 0;
        for(let i of value) {
            sum += i.v;
        }
        temp.set(key, sum);
    }


    const sortedMap = [...temp].sort((a, b) => b[1] - a[1]);

    var answer = [];
    for (let [key, value] of sortedMap) {
        
        const maxCount = Math.min(map.get(key).length, 2);
        for (let i = 0; i < maxCount; i++) {
            answer.push(map.get(key)[i].no);
        }
        
    }
    return answer;
}